ambari-server/src/main/java/org/apache/ambari/server/state/alert/AlertDefinition.java

```diff
@@ -419,4 +428,33 @@ public int hashCode() {
   public String toString() {
     return name;
   }
+
+  /**
+   * Collect the host names from the cluster where the given alert is allowed to run.
+   * A host is able to run an alert if the service component of the alert is installed on that particular host.
+   * In case of AMBARI server alerts or AGGREGATE alerts, an empty host set is returned.
+   * @return matching host names
+   */
+  public Set<String> matchingHosts(Clusters clusters) throws AmbariException {
+    Cluster cluster = clusters.getCluster(clusterId);
+    if (source.getType() == SourceType.AGGREGATE) {
+      return emptySet();
+    }
+    if (AMBARI.name().equals(serviceName)) {
+      return AMBARI_AGENT.name().equals(componentName) ? cluster.getHostNames() : emptySet();
+    }
+    Set<String> matchingHosts = new HashSet<>();
+    for (String host : cluster.getHostNames()) {
+      for (ServiceComponentHost component : cluster.getServiceComponentHosts(host)) {
+        if (belongsTo(component)) {
+          matchingHosts.add(host);
+        }
+      }
+    }
+    return matchingHosts;
+  }
+
+  private boolean belongsTo(ServiceComponentHost component) {
+    return component.getServiceName().equals(serviceName) && component.getServiceComponentName().equals(componentName);
```

Comment:
According to 
https://github.com/mpapirkovskyy/ambari/blob/0907d6119f4e1a0f0ae1add6302cd2ad8fe4d149/ambari-server/src/main/java/org/apache/ambari/server/orm/entities/AlertDefinitionEntity.java#L74
it is possible for componentName to be NULL. However I didn't find such entries on real cluster so not 100% sure.
Can you double check this please?

Reply:
If it's null it won't fail but the give host won't be included. I only see alerts with null component names where sourceType is AGGREGATE and this case is handled separately. Our existing invalidateHosts() method worked similarly so I suppose it's ok.